### ISSUE TITLE
tolist() should have a consistent output type

### ISSUE THREAD

**Issue by yasuiniko** [numpy/numpy#16243]
As mentioned in #11413, the `tolist()` method does not always return a list. The naming of `tolist()` implies that the output will be a list, which can cause issues as in the example below. 

If the recursive part of `tolist()` was rewritten with a helper function, then `tolist()` could have type-consistent behaviour regardless of the number of dimensions of the input array. 

### Reproducing code example:
```python
import numpy as np
array_1D_or_2D = np.arange(3)
for x in array_1D_or_2D:
    len(x.tolist())
```
### Error message:
```python
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: object of type 'int' has no len()
```

**Comment by eric-wieser** [comment]
Note that this is documented behavior:

> If a.ndim is 0, then since the depth of the nested list is 0, it will not be a list at all, but a simple Python scalar.

**Comment by yasuiniko** [comment]
Yes, it is documented. Thank you. I believe the behaviour itself should be changed. Does this need to be marked as a feature request instead?

**Comment by eric-wieser** [comment]
What exactly are you proposing the value of `np.array(42).tolist()` should be instead of the current `42`? Remember it needs to be different from `np.array([42]).tolist()`.

**Comment by yasuiniko** [comment]
Why does the output need to be different? A method called `tolist()` should take the input array and turn it into a list. Even though `np.array(42)` is a scalar, `np.array(42).tolist()` should return a list.

If the reason they have to be different is to enable the recursion, then the recursion can just be moved to a different private function that is called by `tolist()`.

**Comment by eric-wieser** [comment]
They need to be different because of the intent that `array(x.tolist()) == x`.

If you need your list to be 1d, then use `atleast1d(x).tolist()`

**Comment by yasuiniko** [comment]
Ah, it looks like I didn't understand the intended usage. Could the documentation be updated to reflect your answer? I brought the current usage up with my lab and several people were surprised, so updating the documentation might help a lot of people. 

Thanks for the pointer to `atleast_1d()`!

**Comment by eric-wieser** [comment]
To some extent the documentation already mentions that use case. If you can suggest a clearer wording, a pull request would be welcome!

**Comment by yasuiniko** [comment]
I'll see if I can think of something!

**Comment by eric-wieser** [comment]
Out of curiosity, what is your current use case?

**Comment by yasuiniko** [comment]
My use case was that I have to aggregate several numpy arrays into a csv, with each row in the csv containing a single row from each numpy array (as well as some other data from another source). The numpy arrays are usually two-dimensional, so I used `for row in array` to get a one-dimensional array, and then `str(row.tolist())` to get the data for the csv. Sometimes the arrays are one-dimensional, in which case `for row in array` iterates over scalars, so the output from `str(row.tolist())` was not surrounded by "[...]" as I expected. 

It's a bit embarrassing, but I looked back on my code and (surprise surprise) it looks like it's my fault. The dimensionality issue occurred because the original data for each array was a list of lists — but if the inner lists had dimension 1 they were being converted to scalars incorrectly by my code before being sent to numpy.
